[atomic_openshift] New plugin for Atomic/OpenShift

codificat opened the pull request:
Add a new plugin to collect config and runtime information for Atomic Platform / OpenShift Enterprise 3+ to fix #612.

The plugin requires the ignore_case option introduced in #780.

bmr-cymru commented:
pep8...

bmr-cymru commented on sos/plugins/origin.py:
```diff
@@ -0,0 +1,132 @@
+    master_cfg = os.path.join(master_base_dir, "master-config.yaml")
+    node_cfg = os.path.join(node_base_dir, "node-config.yaml")
+    admin_cfg = os.path.join(master_base_dir, "admin.kubeconfig")
+    oc_cmd = "oc --config=%s" % admin_cfg
```
Just as a style thing I think there's a general preference now for having the 'general' Plugin members come first, followed by blocks of plugin-specific instance variables.

It's purely a style/convenience thing but it's handy being able to open up a file, glance below the class line and see the name/profile/package/whatever lists.

I should add this to the Plugin writer's guide..

codificat replied:
Ack, I'll move the name and profile to the top. I need to keep `files` last though as it references instance variables for paths.

bmr-cymru replied:
Tricky. I think we need to do something better here (the idea is to not 'hide' the global plugin properties). Let me think about this some more - if nothing else we can assign a dummy value in the 1st block and then update it later (iirc that's fine in py2/3), so something like:

```python
        plugin_name = "foo"
        files = None # file lists assigned following path set up
```

Or whatever...

bmr-cymru commented on sos/plugins/origin.py:
```diff
@@ -0,0 +1,132 @@
+            self.add_copy_spec([
+                self.master_cfg,
+                os.path.join(self.master_base_dir, "*.crt"),
+            ])
```
I know it's a small bit of code but as this is basically identical to the node case (aside the path value) we could factor it out and save a few lines.

Maybe something like:

```python
def get_origin_config(self, base_dir, cfg):
        self.add_copy_spec([
            os.path.join(base_dir, "*.crt"),
            cfg
        ])

self.get_origin_config(self.master_base_dir, self.master_cfg)
self.get_origin_config(self.node_base_dir, self.node_cfg)
```

OR, this would be more work, but may end up cleaner -  set up the `base_dir` and `config` members according to whether we're a master or node and then these aspects of the collection are completely common and can be moved outside the `is_master/node()` tests. This could be done in a constructor or at the start of the `setup()` method.

codificat replied:
The problem here is that masters can also be nodes - and in fact they are on most deployments. Therefore, when ran from a master we need to collect info for master and node, with their respective different base dirs.

I see your point and I appreciate that this seems to have potential for optimization, but I believe that in this case trying to unify this would make things more confusing in the end.

bmr-cymru replied:
> The problem here is that masters can also be nodes - and in fact they are on most deployments.

OK - thanks - this definitely deserves a mention in a comment in that case.

codificat replied:
The comment is there, line 90, before the `is_node` test... I'll move it (or copy it) before the `is_master` test.

bmr-cymru replied:
Not big / clear enough :-)

I'd suggest putting a large block comment before either of the tests along the lines of:

```python
        #' Master vs. node
        #
        # Openshift Origin/3.x/whatever permits any node to be designated either a master
        # a node, or both (or whatever the rules are). The following methods may all evaluate
        # True on a single instance (at least one must evaluate True if this is an Origin installation)

Feel free to use a docstring if you think that makes more sense although I think a block comment works best as it applies to several methods.

bmr-cymru commented on sos/plugins/origin.py:
```diff
@@ -0,0 +1,132 @@
+    '''
+
+    packages = ('atomic-openshift',)
+
```
Ack, I think this may be our smallest plugin :-)

bmr-cymru commented:
Ping 3.3 is closing very soon - are we ready to go on this?

codificat commented:
Previous feedback added. Also updated after some recent changes to OpenShift (e.g. diagnostics command). Rebased and now using `add_journal`.
Passes smoke test. I would have liked to test more extensively but I believe it's ready to go, at least as a start.
Just a reminder that this PR depends on #718 for `ignore_case` in the output subs.

pmoravec commented:
I reviewed the commit (proper link is https://github.com/codificat/sos/commit/88556f75036a3a1de54983d7917c97b799e07286 not the one github offers above), seems legit :thumbsup:

bmr-cymru commented:
Thanks Pep! I'll try to get this done later today.

codificat commented:
Thanks @pmoravec and @bmr-cymru. I just added a last-minute chage: a way to collect the node's kube config using `oc config view`, which has built-in secret obfuscation (so no need to postproc that).
If you want me to revert back this to the version that Pavel reviewed just let me know.

pmoravec commented:
Changes sound ok to me, I like the "If we fail to find a specific kubeconfig .." part to be there :)

codificat commented:
Updated again: now changed the `postproc` method to not use `ignore_case` which was previously planned in #718 but is not making it. Instead include lower and upper case versions of the target strings.

codificat commented:
This was merged to master in dc37f9f so closing it.
